File: packages/igloo-ts/src/widgets/radio.ts
import { Radio as WitRadio } from 'iced:app/radio@0.1.0';
import type { CallbackId } from 'iced:app/callbacks@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { LineHeight, Shaping, Wrapping } from 'iced:app/text@0.1.0';
import type { Pixels } from 'iced:app/shared@0.1.0';
import { Element, type IntoElement } from '../element.js';
import { pushFixed } from '../callbacks.js';

/**
 * Builder for creating Radio button widgets.
 * A Radio is a circular button representing an alternative in a group.
 *
 * @example
 * ```typescript
 * const radio = Radio.new('Option A', state.selected === 'a', () => ({
 *   type: 'selected',
 *   value: 'a',
 * }));
 * ```
 *
 * @typeParam Msg - The application message type
 */
export class Radio<Msg> implements IntoElement {
  private raw: WitRadio;

  private constructor(label: string, isSelected: boolean, msg: CallbackId) {
    this.raw = new WitRadio(label, isSelected, msg);
  }

  /**
   * Create a new Radio builder.
   * @param onSelect - Handler called when this radio is selected
   */
  static new<Msg>(label: string, isSelected: boolean, onSelect: () => Msg): Radio<Msg> {
    return new Radio(label, isSelected, pushFixed(onSelect()));
  }

  /** Set the radio button size in pixels */
  size(size: Pixels): this {
    this.raw.size(size);
    return this;
  }

  /** Set the width */
  width(width: Length): this {
    this.raw.width(width);
    return this;
  }

  /** Set the spacing between radio button and label */
  spacing(spacing: Pixels): this {
    this.raw.spacing(spacing);
    return this;
  }

  /** Set the text size */
  textSize(size: Pixels): this {
    this.raw.textSize(size);
    return this;
  }

  /** Set the text line height */
  textLineHeight(lineHeight: LineHeight): this {
    this.raw.textLineHeight(lineHeight);
    return this;
  }

  /** Set the text wrapping */
  textWrapping(wrapping: Wrapping): this {
    this.raw.textWrapping(wrapping);
    return this;
  }

  /** Set the text shaping */
  textShaping(shaping: Shaping): this {
    this.raw.textShaping(shaping);
    return this;
  }

  /** Convert to Element */
  intoElement(): Element {
    return new Element(WitRadio.intoElement(this.raw));
  }
}

File: crates/igloo_guest/src/widgets/radio.rs
use iced_core::{Length, Pixels, text};

use crate::Element;
use crate::bindings::iced::app::widgets::{Node, RadioNode};

/// A circular button representing an alternative.
pub struct Radio<Message> {
    label: String,
    is_selected: bool,
    on_select: Message,
    size: Option<f32>,
    width: Option<Length>,
    spacing: Option<f32>,
    text_size: Option<f32>,
    text_line_height: Option<text::LineHeight>,
    text_wrapping: Option<text::Wrapping>,
    text_shaping: Option<text::Shaping>,
}

impl<Message: 'static> Radio<Message> {
    /// Creates a new [`Radio`] with the given label and value.
    pub fn new<F, V>(label: impl Into<String>, value: V, selected: Option<V>, f: F) -> Self
    where
        V: Eq + Copy,
        F: FnOnce(V) -> Message,
    {
        Self {
            label: label.into(),
            is_selected: Some(value) == selected,
            on_select: f(value),
            size: None,
            width: None,
            spacing: None,
            text_size: None,
            text_line_height: None,
            text_wrapping: None,
            text_shaping: None,
        }
    }

    pub fn size(mut self, size: impl Into<Pixels>) -> Self {
        self.size = Some(size.into().0);
        self
    }

    pub fn width(mut self, width: impl Into<Length>) -> Self {
        self.width = Some(width.into());
        self
    }

    pub fn spacing(mut self, spacing: impl Into<Pixels>) -> Self {
        self.spacing = Some(spacing.into().0);
        self
    }

    pub fn text_size(mut self, size: impl Into<Pixels>) -> Self {
        self.text_size = Some(size.into().0);
        self
    }

    pub fn text_line_height(mut self, line_height: impl Into<text::LineHeight>) -> Self {
        self.text_line_height = Some(line_height.into());
        self
    }

    pub fn text_wrapping(mut self, wrapping: impl Into<text::Wrapping>) -> Self {
        self.text_wrapping = Some(wrapping.into());
        self
    }

    pub fn text_shaping(mut self, shaping: impl Into<text::Shaping>) -> Self {
        self.text_shaping = Some(shaping.into());
        self
    }
}

impl<Message: 'static> From<Radio<Message>> for Element<Message> {
    fn from(radio: Radio<Message>) -> Self {
        Element::new(move |realize, arena| {
            let node = RadioNode {
                label: radio.label,
                is_selected: radio.is_selected,
                on_select: realize.fixed(radio.on_select),
                size: radio.size,
                width: radio.width.map(Into::into),
                spacing: radio.spacing,
                text_size: radio.text_size,
                text_line_height: radio.text_line_height.map(Into::into),
                text_wrapping: radio.text_wrapping.map(Into::into),
                text_shaping: radio.text_shaping.map(Into::into),
            };
            arena.push(Node::Radio(node))
        })
    }
}
